Add arbos customize docs page (#1563)
OffchainLabs/arbitrum-docs: arbitrum-docs/launch-orbit-chain/how-tos/customize-arbos.mdx

---
title: "How to customize ArbOS on your Orbit chain"
description: "Learn how to customize your ArbOS version"
author: jasonwan
sidebar_position: 7
content_type: how-to
---

import PublicPreviewBannerPartial from '../partials/_orbit-public-preview-banner-partial.md';

<PublicPreviewBannerPartial />

> 

Reviewer: ```suggestion
```caution Customizations may not be for everyone
Being able to customize and control every aspect of your chain is one of the key benefits of building with the Arbitrum Orbit tech stack. Performing these customizations, however, is complex and may not be tenable for every team (depending on the type of changes your team is trying to make). 

Due to the complex nature of ArbOS customizations, we strongly recommend being very intentional with your chain's customizations and budget for extra resources to perform the short term work, but also to manage the medium and long term maintenance and security risks. Despite our desire to help every team with their customizations, the Offchain Labs team has limited capacity to review and audit every team's customizations but we will do our best, alongside your RaaS teams, to support you in any way we can.
```
```

        ensure(state.SetNewMyNumber(${random number}))
    //....
} 
```

Here, we will ensure that the initial value of $(an random number) after ArbOS is upgraded to V21.

:::caution
It should be noted that when you initialize the state (initial code is in [customize precompile Option 5](./customize-precompile.mdx#option-5-call-and-modify-state)), you need to initialize it to 0 or null value first to avoid potential blockchain reorg.
:::

And, please make sure that your program cannot call the `state.SetNewMyNumber` or other function might change the value of `myNumber` before ArbOS V21. 
To prevent this, if you are using an external call to the precompile contract to change the value, you can refer to [point 1](./customize-arbos.mdx#1-add-a-new-method-to-exsiting-precompile-on-a-specific-arbos-version) or
[point 2](./customize-arbos.mdx#2-create-a-new-precompile-contract-on-a-specific-arbos-version) to set the activation time of the precompile contract method.
If your nitro code needs to call this method to change the state, you can continue reading [point 4](./customize-arbos.mdx#4-any-changes-in-the-stf-logic-that-will-affect-the-final-execution-result).

> ### 4. Any changes in the stf logic that will affect the final execution result

Reviewer: ```suggestion
### 4. Any changes in the STF logic that will affect the final execution result
```

---
title: "How to customize ArbOS on your Orbit chain"
description: "Learn how to customize your ArbOS version"
author: jasonwan
sidebar_position: 7
content_type: how-to
---

import PublicPreviewBannerPartial from '../partials/_orbit-public-preview-banner-partial.md';
import CustomizeCautionBannerPartial from '../partials/_orbit-customize-caution-banner-partial.md'

<PublicPreviewBannerPartial />

<CustomizeCautionBannerPartial />

## In which case you need to customize ArbOS upgrade

> First, when you make changes to your nitro code and it may affect State Transition Function (You can refer to [Customize STF](./customize-stf.mdx#Preface)), 

Reviewer: ```suggestion
1. When you want to make changes to your Nitro code that affect the State Transition Function, or STF (you may refer to the [Customize STF docs](./customize-stf.mdx#Preface)), and
```

Second, if your Orbit Chain is not started for the first time but the changes are made while it is running.

If your changes meet both those 2 points, then you need to customize the ArbOS upgrade.

Also, if you added some changes to the chain and want to upgrade them at some time in the future. You should also take it into ArbOS upgrade process.

## Where should I insert ArbOS Upgrade related code?

> Here, we will provide you with 4 examples that need to add ArbOS Upgrade-related code:

Reviewer: ```suggestion
Below, you will find 4 examples of ArbOS-related code changes and, generally, how to make them:
```

In this way, ArbHi and all its methods will be activated after the ArbOS version you set.

### 3. Create a new ArbOS state on a specific ArbOS version

After you add a new state `myNumber` according to the guide in [customize precompile Option 5](./customize-precompile.mdx#option-5-call-and-modify-state), you also need to rewrite [UpgradeArbosVersion][arbosstate_updatearbosversion_method_link] in [arbosstate.go][arbosstate_file_link]:
Add your expected ArbOS version to the switch case statement of nextArbosVersion. Here we will take ArbOS V21 as an example:
```go
ensure := func(err error) {
    if err != nil {
        message := fmt.Sprintf(
            "Failed to upgrade ArbOS version %v to version %v: %v",
            state.arbosVersion, state.arbosVersion+1, err,
        )
        panic(message)
    }
}
nextArbosVersion := state.arbosVersion + 1
switch nextArbosVersion {
    case 1:
        //.....
    case 2:
        //....
    //.....
    case 21:
        // Set your new ArbOS state value here
        ensure(state.SetNewMyNumber(${random number}))
    //....
} 
```

Here, we will ensure that the initial value of $(a random number) after ArbOS is upgraded to V21.

:::caution

> It should be noted that when you initialize the state (initial code is in [customize precompile Option 5](./customize-precompile.mdx#option-5-call-and-modify-state)), you need to initialize it to 0 or null value first to avoid potential blockchain reorg.

Reviewer: ```suggestion
It should be noted that when you initialize the state (initial code is in [customize precompile Option 5](./customize-precompile.mdx#option-5-call-and-modify-state)), you need to initialize it to 0 or null value first to avoid a potential re-org on your chain.
```

:::

> Also, please make sure that your program cannot call the `state.SetNewMyNumber` or other functions may change the value of `myNumber` before ArbOS V21. 

Reviewer: ```suggestion
Also, please make sure your program cannot call the `state.SetNewMyNumber` or other functions may change the value of `myNumber` before ArbOS V21. 
```

### 4. Any changes in the STF logic that will affect the final execution result
If you change the logic in STF, it will cause the execution result of the transaction to be different, 
you need to keep the original execution logic and put the new logic into another branch.
You can use an `if else` statement to control it.

For example, we change the `SayHi` return after upgrading the ArbOS version:
```go
// The method in ArbHi precompile
func (con *ArbHi) SayHi(c ctx, evm mech) (string, error) {
    if p.state.ArbOSVersion() >= ${The arbos you want to upgrade to} {
        // Your new logic code
        return "hi, new ArbOS version", nil
    } else {
        // The old logic code needs to be kept
        return "hi", nil
    }
}
```

In the above code, we use precompiles as an example. Some logic might also affect the STF, such as the methods in [block_process.go][block_processor_file_link], [internal_tx.go][internal_tx_file_link], [tx_processor.go][tx_processor_file_link] and so on.
You also need to use those ArbOS control methods for different logic versions.

> ## Schedule ArbOS upgrade

Reviewer: Shouldn't the scheduling of the ArbOS upgrade always happen *after* the WASM module root is set? https://docs.arbitrum.io/launch-orbit-chain/how-tos/arbos-upgrade

---
title: "How to customize ArbOS on your Orbit chain"
description: "Learn how to customize your ArbOS version"
author: jasonwan
sidebar_position: 7
content_type: how-to
---

import PublicPreviewBannerPartial from '../partials/_orbit-public-preview-banner-partial.md';
import CustomizeCautionBannerPartial from '../partials/_orbit-customize-caution-banner-partial.md'

<PublicPreviewBannerPartial />

<CustomizeCautionBannerPartial />

## Cases where you may want to consider customizing your own ArbOS upgrade

1. When you want to make changes to your Nitro code that affect the State Transition Function, or STF (you may refer to the [Customize STF docs](./customize-stf.mdx#Preface)), and

2. If your desired changes need to be made to a live and operational Orbit chain

> If your changes meet both those 2 points, then you need to customize the ArbOS upgrade.

Reviewer: ```suggestion
If your changes meet both those 2 points, then you will need a custom ArbOS upgrade.
```

If your nitro code needs to call this method to change the state, you can continue reading [point 4](./customize-arbos.mdx#4-any-changes-in-the-stf-logic-that-will-affect-the-final-execution-result).

### 4. Any changes in the STF logic that will affect the final execution result
If you change the logic in STF, it will cause the execution result of the transaction to be different, 
you need to keep the original execution logic and put the new logic into another branch.
You can use an `if else` statement to control it.

For example, we change the `SayHi` return after upgrading the ArbOS version:
```go
// The method in ArbHi precompile
func (con *ArbHi) SayHi(c ctx, evm mech) (string, error) {
    if p.state.ArbOSVersion() >= ${The arbos you want to upgrade to} {
        // Your new logic code
        return "hi, new ArbOS version", nil
    } else {
        // The old logic code needs to be kept
        return "hi", nil
    }
}
```

In the above code, we use precompiles as an example. Some logic might also affect the STF, such as the methods in [block_process.go][block_processor_file_link], [internal_tx.go][internal_tx_file_link], [tx_processor.go][tx_processor_file_link] and so on.

> You also need to use those ArbOS control methods for different logic versions.

Reviewer: Are you referring to the `.go` methods? 
```suggestion
The aforementioned ArbOS control methods will be needed for interacting with different versions of the STF logic.
```

Does this match what you're trying to say?

:::tip Backward compatibility

Wasm module roots are backward compatible, so upgrading them before an ArbOS version upgrade will not disrupt your chain's functionality.

:::

## Upgrade WASM Module Root on the parent chain

> Afterward, since Arbitrum might need to execute the validation and fraud proofs on the parent chain, you need to update the Wasm Module Root recorded on the parent chain.

Reviewer: ```suggestion
After you have made your custom ArbOS changes, you will need to update the WASM Module root recorded on the parent chain. This is because an Arbitrum chain may execute the validation and fraud proofs on the parent chain.
```

What does it mean to "execute validation on the parent chain"?

Please continue reading [customize stf](./customize-stf.mdx#step-3-run-the-node-without-fraud-proofs) for follow-up operations.

## Schedule ArbOS upgrade

> After you add ArbOS version control to the nitro code, you can update ArbOS. You can refer to the document [ArbOS upgrade](/launch-orbit-chain/how-tos/arbos-upgrade) to upgrade.

Reviewer: ```suggestion
After you add ArbOS version control to the Nitro code, you can update ArbOS. You can refer to the document [ArbOS upgrade](/launch-orbit-chain/how-tos/arbos-upgrade) to upgrade. It is recommended that teams who opt to version control their custom ArbOS version choose an ArbOS version number that builds on top of the canonical releases shipped by Offchain Labs. As an example, if your team is customizing ArbOS 31, we recommend versioning your ArbOS version as ArbOS 32. 
```
